controller/grpc/controller.go

```diff
@@ -0,0 +1,1179 @@
+		}
+		releases = append(releases, r)
+	}
+
+	stream.Send(&protobuf.StreamReleasesResponse{
+		Releases:      releases,
+		NextPageToken: nextPageToken.String(),
+		PageComplete:  true,
+	})
+
+	if unary {
+		return nil
+	}
+
+	unmarshalRelease := func(event *ct.Event) (*protobuf.Release, error) {
+		var ctRelease *ct.Release
+		if err := json.Unmarshal(event.Data, &ctRelease); err != nil {
+			return nil, protobuf.NewError(err, err.Error())
+		}
+		return protobuf.NewRelease(ctRelease), nil
+	}
+
+	maybeAcceptRelease := func(event *ct.Event) (release *protobuf.Release, accepted bool) {
+		r, err := unmarshalRelease(event)
+		if err != nil {
+			s.logger.Error("error unmarshalling event", "rpcMethod", "StreamReleases", "event_id", event.ID, "error", err)
+			return
+		}
+
+		if !protobuf.MatchLabelFilters(r.Labels, req.GetLabelFilters()) {
+			return
+		}
+
+		return r, true
+	}
+
+	// stream new events as they are created
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		var currID int64
+		for {
+			event, ok := <-sub.Events
+			if !ok {
+				break
+			}
+
+			// avoid overlap between list and stream
+			if event.ID <= currID {
+				continue
+			}
+			currID = event.ID
+
+			if release, ok := maybeAcceptRelease(event); ok {
+				stream.Send(&protobuf.StreamReleasesResponse{
+					Releases: []*protobuf.Release{release},
+				})
+			}
+		}
+	}()
+	wg.Wait()
+
+	return maybeError(sub.Err)
+}
+
+func (s *server) CreateRelease(ctx context.Context, req *protobuf.CreateReleaseRequest) (*protobuf.Release, error) {
+	ctRelease := req.Release.ControllerType()
+	ctRelease.AppID = protobuf.ParseIDFromName(req.Parent, "apps")
+	if err := s.releaseRepo.Add(ctRelease); err != nil {
+		return nil, protobuf.NewError(err, err.Error())
+	}
+	return protobuf.NewRelease(ctRelease), nil
+}
+
+func (s *server) listDeployments(req *protobuf.StreamDeploymentsRequest) ([]*protobuf.ExpandedDeployment, *data.PageToken, error) {
+	pageToken, err := data.ParsePageToken(req.PageToken)
+	if err != nil {
+		return nil, nil, err
+	}
+	if req.PageSize > 0 {
+		pageToken.Size = int(req.PageSize)
+	}
+	statusFilters := make([]string, 0, len(req.StatusFilters))
+	for _, status := range req.StatusFilters {
+		statusFilters = append(statusFilters, status.ControllerType())
+	}
+	ctExpandedDeployments, nextPageToken, err := s.deploymentRepo.ListPage(data.ListDeploymentOptions{
+		PageToken:     *pageToken,
+		AppIDs:        protobuf.ParseIDsFromNameFilters(req.NameFilters, "apps"),
+		DeploymentIDs: protobuf.ParseIDsFromNameFilters(req.NameFilters, "deployments"),
+		StatusFilters: statusFilters,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	deployments := make([]*protobuf.ExpandedDeployment, 0, len(ctExpandedDeployments))
+	for _, d := range ctExpandedDeployments {
+		deployments = append(deployments, protobuf.NewExpandedDeployment(d))
+	}
+
+	var filtered []*protobuf.ExpandedDeployment
+	typeMatcher := protobuf.NewReleaseTypeMatcher(req.TypeFilters)
```

Comment:
This should be in the SQL query to avoid empty responses where there are results. I think it makes sense to add a `type` column to `deployments` to facilitate this.